sdk/core/Azure.Core/src/DynamicData/AllowListConverterFactory.cs

```diff
@@ -0,0 +1,196 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using System.Text.Json;
+using System.Text.Json.Serialization;
+using Azure.Core.Dynamic;
+
+namespace Azure.Core.Json
+{
+    internal class AllowListConverterFactory : JsonConverterFactory
+    {
+        public static readonly AllowListConverterFactory Default = new();
+
+        public override bool CanConvert(Type typeToConvert)
+        {
+            return !IsAllowedType(typeToConvert);
+        }
+
+        public static bool IsAllowedType(Type type)
+        {
+            if (IsAllowedKnownType(type))
+            {
+                return true;
+            }
+
+            return IsAllowedPoco(type, new List<Type>());
+        }
+
+        private static bool IsAllowedPoco(Type type, List<Type> ancestorTypes)
```

Comment:
Would it make sense to make ancestorTypes a HashSet to avoid O(N) lookups?

Reply:
The goal here is to break cycles in type checks - Krzysztof called out that we should [avoid an allocation](https://github.com/Azure/azure-sdk-for-net/pull/36791#discussion_r1228362083), so I'll look into this.

Reply:
This has been changed to an array that is periodically reallocated.